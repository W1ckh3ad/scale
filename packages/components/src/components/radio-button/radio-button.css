.input {
  position: relative;
}
.input .input__helper-text {
  font-weight: var(--scl-font-weight-bold);
}

.input .input__helper-text {
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-default);
  padding-left: var(--scl-spacing-12);
  font-size: var(--scl-font-size-12);
  line-height: var(--scl-font-line-height-133);
  color: var(--scl-color-blue-70);
}
.input .input__meta {
  display: flex;
  justify-content: space-between;
}

.input--status-error .input__helper-text {
  color: var(--scl-color-text-error);
}

.input--type-radio {
  display: flex;
  flex-wrap: wrap;
  align-items: center;
}
.input--type-radio .input__meta {
  width: 100%;
}
.input--type-radio .input__helper-text {
  margin-top: var(--scl-spacing-4);
  padding-left: var(--scl-spacing-24);
}
.input--type-radio label {
  color: var(--scl-color-text-default);
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-default);
  font-weight: var(--scl-font-weight-medium);
}
.input--type-radio input {
  width: var(--scl-spacing-16);
  height: var(--scl-spacing-16);
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-default);
  border-radius: 50%;
  -webkit-appearance: none;
  background-color: #fff;
  border: var(--scl-spacing-1) solid var(--scl-color-text-default);
  margin: 0 var(--scl-spacing-8) 0 0;
}
.input--type-radio input:focus {
  outline: none;
  box-shadow: 0 0 0 var(--scl-spacing-2) var(--scl-color-focus);
}
.input--type-radio:hover input:not(:checked):not([disabled]) {
  box-shadow: none;
  border-color: var(--scl-color-primary-hover, #f90984);
}
.input--type-radio:hover input:not(:checked):not([disabled]) ~ label {
  color: var(--scl-color-primary-hover, #f90984);
}
.input--type-radio input:active {
  border: var(--scl-spacing-8) solid var(--scl-color-primary-active);
}
.input--type-radio input:not(:checked):not([disabled]):active ~ label {
  color: var(--scl-color-primary-active);
}
.input--type-radio input:disabled {
  border: var(--scl-spacing-1) solid var(--scl-color-background-disabled);
}
.input--type-radio input:disabled ~ label {
  color: var(--scl-color-grey-30);
}
.input--type-radio.input--status-error input {
  border: var(--scl-spacing-2) solid var(--scl-color-background-error);
}
.input--type-radio input:checked {
  border: calc(0.5 * var(--scl-spacing-12)) solid var(--scl-color-primary);
}
.input--type-radio input:checked:active {
  border: calc(0.5 * var(--scl-spacing-12)) solid
    var(--scl-color-primary-active, #cb0068);
}
.input--type-radio input:checked:disabled {
  background: var(--scl-color-grey-30);
  border: calc(0.5 * var(--scl-spacing-12)) solid
    var(--scl-color-background-disabled);
}
.input--type-radio input:checked:disabled ~ label {
  color: var(--scl-color-grey-30);
}
.input--disabled label,
.input--disabled .input__label,
.input--disabled input,
.input--disabled .input__radio,
.input--disabled .input__helper-text {
  cursor: not-allowed;
  border-color: var(--scl-color-background-disabled);
  color: var(--scl-color-background-disabled);
  background: var(--scl-color-white);
}
