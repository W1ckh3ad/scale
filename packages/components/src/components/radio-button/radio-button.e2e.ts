import { newE2EPage } from '@stencil/core/testing';

describe('scale-radio-button', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent('<scale-radio-button/>');
    const element = await page.find('scale-radio-button');
    expect(element).toHaveClass('hydrated');
  });
});
