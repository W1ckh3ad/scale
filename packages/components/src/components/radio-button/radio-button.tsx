import {
  Component,
  Event,
  EventEmitter,
  h,
  Host,
  Prop,
  Watch,
} from '@stencil/core';
import classNames from 'classnames';

interface InputChangeEventDetail {
  value: string | number | boolean | undefined | null;
}

let i = 0;

@Component({
  tag: 'scale-radio-button',
  styleUrl: './radio-button.css',
  shadow: false,
})
export class RadioButton {
  /** (optional) Input name */
  @Prop() name?: string = '';
  /** (optional) Input label */
  @Prop() label: string = '';
  /** (optional) Input helper text */
  @Prop() helperText?: string = '';
  /** (optional) Input status */
  @Prop() status?: string = '';
  /** (optional) Input disabled */
  @Prop() disabled?: boolean;
  /** (optional) Active switch */
  @Prop({ reflect: true }) checked?: boolean = false;
  /** (optional) Input value */
  @Prop({ mutable: true }) value?: string | number | null = '';
  /** (optional) Input checkbox id */
  @Prop() inputId?: string;
  /** (optional) Injected CSS styles */
  @Prop() styles?: string;

  /** Emitted when the value has changed. */
  @Event() scaleChange!: EventEmitter<InputChangeEventDetail>;

  componentWillLoad() {
    if (this.inputId == null) {
      this.inputId = 'input-' + i++;
    }
  }

  // We're not watching `value` like we used to
  // because we get unwanted `scaleChange` events
  // because how we keep this.value up-to-date for type="select"
  // `this.value = selectedValue`
  emitChange() {
    this.scaleChange.emit({
      value: this.value == null ? this.value : this.value.toString(),
    });
  }

  @Watch('checked')
  checkedChanged() {
    this.scaleChange.emit({ value: this.checked });
  }

  // Handle checkbox/radio change (click on label)
  handleCheckChange = event => {
    this.checked = event.target.checked;
  };

  render() {
    const ariaInvalidAttr =
      this.status === 'error' ? { 'aria-invalid': true } : {};
    const helperTextId = `helper-message-${i}`;
    const ariaDescribedByAttr = { 'aria-describedBy': helperTextId };

    return (
      <Host>
        <div class={this.getCssClassMap()}>
          <input
            type="radio"
            name={this.name}
            id={this.inputId}
            onChange={this.handleCheckChange}
            value={this.value}
            checked={this.checked}
            disabled={this.disabled}
            {...ariaInvalidAttr}
            {...ariaDescribedByAttr}
          />
          <label htmlFor={this.inputId}>{this.label}</label>
          {!!this.helperText && (
            <div
              class="input__meta"
              id={helperTextId}
              aria-live="polite"
              aria-relevant="additions removals"
            >
              <div class="input__helper-text">{this.helperText}</div>
            </div>
          )}
        </div>
      </Host>
    );
  }

  getCssClassMap() {
    return classNames(
      'input',
      'input--type-radio',
      this.checked && `input--checked`,
      this.disabled && `input--disabled`,
      this.status && `input--status-${this.status}`
    );
  }
}
