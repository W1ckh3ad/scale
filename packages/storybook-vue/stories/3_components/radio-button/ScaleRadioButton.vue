<template>
  <scale-input
    :name="name"
    :status="status"
    :type="type"
    :label="label"
    :helper-text="helperText"
    :disabled="disabled"
    :checked="checked"
    :value="value"
    :input-id="inputId"
    @scaleChange="scaleChange"
  >
    <slot></slot>
  </scale-input>
</template>

<script>
import { action } from "@storybook/addon-actions";

export default {
  props: {
    status: { type: String },
    name: { type: String },
    type: { type: String },
    label: { type: String },
    helperText: { type: String },
    disabled: { type: Boolean },
    checked: { type: Boolean },
    value: { type: String },
    inputId: { type: String },
    styles: { type: String }
  },
  methods: {
    scaleChange($event) {
      action("scaleChange");
      this.$emit("scaleChange", $event);
    }
  }
};
</script>
