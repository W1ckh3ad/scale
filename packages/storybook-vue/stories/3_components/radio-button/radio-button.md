# Radio Button

With radio buttons, users select one option from a group of options in a form.

## General

Radio buttons represent a group of mutually exclusive choices.

![Image Name](/assets/3_components/radio-button/image-20200810094044874.png)

### When  to Use

Whenever only one selection of a group is allowed, use the radio button component.

### When Not to Use

Whenever the user should be able to select several options at once, use checkboxes instead.

## Elements

### Label

Clearly label your radio buttons so users know what option they select e.g. "daily / weekly / monthly".

## Best Practice

### Default Selection

On option within a set of radio buttons should be selected by default when the page is loaded. Never display a group of radio buttons without a default selection.

## Related Components

Checkbox, Switch, Select Box

## Code
