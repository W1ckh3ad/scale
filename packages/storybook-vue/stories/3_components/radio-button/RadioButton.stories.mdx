import { Meta, Story, Props, Preview, Description } from "@storybook/addon-docs/blocks";
import ScaleRadioButton from "./ScaleRadioButton.vue";
import description from "./radio-button.md"
import { action } from "@storybook/addon-actions";
export const ICON = "M16.471,8.321 L12.613594,0.502886112 L8.755,8.321 L0.126435258,9.57533801 L6.37,15.66 L4.89610548,24.2548735 L12.613,20.197 L20.3310826,24.2548735 L18.856,15.66 L25.1007528,9.57533801 L16.471,8.321 Z M9.41891527,9.23556376 L12.613,2.761 L15.8082728,9.23556376 L22.951,10.273 L17.7826928,15.3122039 L19.003,22.427 L12.613594,19.067774 L6.223,22.427 L7.44449521,15.3122039 L2.275,10.273 L9.41891527,9.23556376 Z";

<Meta
  title="Components/RadioButton"
  component={ScaleRadioButton}
  argTypes={{}}
/>

export const Template = (args, { argTypes }) => ({
  components: { ScaleRadioButton },
  props: {
    label: String,
    ...ScaleRadioButton.props
  },
  template: `
    <scale-input
      :name="name"
      :type="type"
      :label="label"
      :disabled="disabled"
      :pre-checked="preChecked"
      :value="value"
      :input-id="inputId"
      @scaleChange="scaleChange"
    ></scale-input>
  `,
  methods: {
    scaleChange: action("scaleChange"),
  }
})

<Description>{description}</Description>

<Preview withSource="none">
  <Story
    name="Standard"
    args={{
      label: "Radio",
      type: "radio"
    }}
  >
    {Template.bind({})}
  </Story>
</Preview>

<Props story="Standard" />


## Pre-checked

<Preview withSource="none">
  <Story
    name="Pre-checked"
    args={{
      label: "Pre-checked",
      type: "radio",
      preChecked: true
    }}
  >
    {Template.bind({})}
  </Story>
</Preview>

```html
<scale-input type="radio" pre-checked></scale-input>
```


## Disabled

<Preview withSource="none">
  <Story
    name="Disabled"
    args={{
      label: "Disabled",
      type: "radio",
      disabled: true
    }}
  >
    {Template.bind({})}
  </Story>
</Preview>

```html
<scale-input type="radio" disabled></scale-input>
```
