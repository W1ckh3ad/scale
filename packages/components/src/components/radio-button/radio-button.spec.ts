import { newSpecPage } from '@stencil/core/testing';
import { RadioButton } from './radio-button';

describe('RadioButton', () => {
  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [RadioButton],
      html: `<scale-radio-button></scale-radio-button>`,
    });
    expect(page.root).toMatchSnapshot();
  });
});
