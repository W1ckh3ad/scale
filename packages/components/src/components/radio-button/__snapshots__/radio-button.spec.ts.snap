// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton should match snapshot 1`] = `
<scale-radio-button>
  <div class="radio-button">
    <input aria-describedby="helper-message-1" id="input-0" type="radio" value="">
    <label htmlfor="input-0"></label>
  </div>
</scale-radio-button>
`;
